chan.h

```diff
@@ -11,45 +11,71 @@ using std::mutex;
 using std::queue;
 using std::unique_lock;
 
-template <typename T>
-class Chan {
-    queue<T> que;
-    mutable mutex mtx;
-    condition_variable cond;
+template <typename T> class Chan {
+  // buffered queue properties
+  queue<T> que;
+  mutable mutex mtx;
+  condition_variable r_cond;
+  condition_variable w_cond;
 
-    const size_t quota;
-    size_t passed;
+  int r_waiting;
+  int w_waiting;
+  const size_t capacity;
+  size_t passed;
 
 public:
-    Chan(size_t quota_) : quota(quota_), passed(0) {}
```

Comment:
What we need:
Chan(size_t quota_, size_t capacity_) : quota(quota_), capacity(capacity_)
When passed == quota, Pop() always return false, means this channel was closed.
When queue.size() == capacity, Push() should wait, till some one call Pop().